src/libutil/cf.c

```diff
@@ -210,6 +215,121 @@ bool cf_array_contains (const cf_t *cf, const char *str)
     return false;
 }
 
+static bool parent_dir_is_secure (cf_t *cf,
+                                  const char *filename,
+                                  struct cf_error *error)
+{
+    struct stat st;
+    char buf [PATH_MAX + 1];
+    char *dir;
+
+    if (strlcpy (buf, filename, sizeof (buf)) >= sizeof (buf)
+        || !(dir = dirname (buf))) {
+        errno = ENAMETOOLONG;
+        errprintf (error, filename, -1, "Unable to get dirname");
+        return false;
+    }
+    if (lstat (dir, &st) < 0) {
+        errprintf (error, filename, -1, "Unable to stat parent directory");
+        return false;
+    }
+    if (!S_ISDIR (st.st_mode)) {
+        errprintf (error, filename, -1,
+                   "Unable to check parent directory. Unexpected file type");
+        errno = EINVAL;
+        return false;
+    }
+    if ((st.st_uid != 0)
+        && (st.st_uid != geteuid ())) {
+        errprintf (error, filename, -1,
+                   "Invalid ownership on parent directory");
+        errno = EINVAL;
+        return false;
+    }
+    if (st.st_gid != 0
+        && st.st_gid != getegid ()
+        && (st.st_mode & S_IWGRP)
+        && !(st.st_mode & S_ISVTX)) {
+        errprintf (error, filename, -1,
+                   "parent directory is group-writeable without sticky bit");
+        errno = EINVAL;
+        return false;
+    }
+    if ((st.st_mode & S_IWOTH)
+        && !(st.st_mode & S_ISVTX)) {
+        errprintf (error, filename, -1,
+                   "parent directory is world-writeable without sticky bit");
+        errno = EINVAL;
+        return false;
+    }
+    errno = 0;
+    return true;
+}
+
+static bool filename_is_secure (cf_t *cf,
+                                const char *filename,
+                                struct cf_error *error)
+{
+    int is_symlink = 0;
+    struct stat st;
+
+    if ((filename == NULL) || (*filename == '\0')) {
+        errprintf (error, "", 0, "Filename not defined");
+        return false;
+    }
+    if ((lstat (filename, &st) == 0)
+        && S_ISLNK (st.st_mode) == 1)
+        is_symlink = 1;
+    if (stat (filename, &st) < 0) {
+        errprintf (error, filename, -1, "Failed to find config file");
+        return false;
+    }
+    if (!S_ISREG (st.st_mode)) {
+        errprintf (error, filename, -1, "File is not a regular file");
+        errno = EINVAL;
+        return false;
+    }
+    if (is_symlink) {
+        errprintf (error, filename, -1, "Refusing to open symbolic link");
+        errno = EINVAL;
+        return false;
+    }
```

Comment:
I take it if it's a symlink to a special file, you wanted to call that out rather than just bail on a symlink no matter what it points to?

Edit: oh when I came to the unit tests I now see why this makes things simpler.

Reply:
Yeah, and TBH I just took the example from munged keyfile tests. Perhaps @dun could comment on what the best approach is here. In the case of flux-security config files, it may actually be convenient to allow symlinks as long as the target also has the correct permissions/ownership?

Reply:
munged checks the permissions and ownership of its directories all the way up to the root. Since a symlink could point outside of this directory subtree, munged simply disallows symlinks. See:

https://cwe.mitre.org/data/definitions/61.html
https://cwe.mitre.org/data/definitions/59.html

If you're going to allow symlinks, I'd recommend the symlink target should have the same ownership and permission checks as the original path (i.e., you'd need to, at least, also check the ownership and permissions of the target directory).

Reply:
Thanks for the background @dun! That is really helpful. I think at this point I'd like to err on the side of being overly cautious and just disallow symlinks when ownership and permission checks are enabled. If this becomes onerous in the future, then this requirement could potentially be relaxed.

Another mitigating factor here is that the cf file and permission checks are only enabled if the process is running with elevated privileges or setuid (euid != uid), so normal users should be able to use symlinks to their heart's content. :heart_eyes: :link: